state/indexer/sink/psql/psql.go

```diff
@@ -55,116 +56,182 @@ func (es *EventSink) DB() *sql.DB { return es.store }
-		sqlStmt, h.ResultBeginBlock.Events, types.EventTypeBeginBlock, h.Header.Height, ts, es.chainID)
+// runInTransaction executes query in a fresh database transaction.
+// If query reports an error, the transaction is rolled back and the
+// error from query is reported to the caller.
+// Otherwise, the result of committing the transaction is returned.
+func runInTransaction(db *sql.DB, query func(*sql.Tx) error) error {
+	tx, err := db.Begin()
 	if err != nil {
 		return err
 	}
-
-	// index end_block events
-	sqlStmt, err = indexBlockEvents(
-		sqlStmt, h.ResultEndBlock.Events, types.EventTypeEndBlock, h.Header.Height, ts, es.chainID)
-	if err != nil {
+	if err := query(tx); err != nil {
+		_ = tx.Rollback() // report the initial error, not the rollback
 		return err
 	}
+	return tx.Commit()
+}
 
-	_, err = sqlStmt.RunWith(es.store).Exec()
-	return err
+// queryWithID executes the specified SQL query with the given arguments,
+// expecting a single-row, single-column result containing an ID. If the query
+// succeeds, the ID from the result is returned.
+func queryWithID(tx *sql.Tx, query string, args ...interface{}) (uint32, error) {
+	var id uint32
+	if err := tx.QueryRow(query, args...).Scan(&id); err != nil {
+		return 0, err
+	}
+	return id, nil
 }
 
-func (es *EventSink) IndexTxEvents(txr []*abci.TxResult) error {
-	// index the tx result
-	var txid uint32
-	sqlStmtTxResult := sq.
-		Insert(TableResultTx).
-		Columns("tx_result", "created_at").
-		PlaceholderFormat(sq.Dollar).
-		RunWith(es.store).
-		Suffix("ON CONFLICT (tx_result)").
-		Suffix("DO NOTHING").
-		Suffix("RETURNING \"id\"")
-
-	sqlStmtEvents := sq.
-		Insert(TableEventTx).
-		Columns("key", "value", "height", "hash", "tx_result_id", "created_at", "chain_id").
-		PlaceholderFormat(sq.Dollar).
-		Suffix("ON CONFLICT (key,hash)").
-		Suffix("DO NOTHING")
-
-	ts := time.Now()
-	for _, tx := range txr {
-		txBz, err := proto.Marshal(tx)
-		if err != nil {
-			return err
-		}
+// insertEvents inserts a slice of events and any indexed attributes of those
+// events into the database associated with tx.
+//
+// If txID > 0, the event is attributed to the transaction with that ID;
+// otherwise it is recorded as a block event.
+func insertEvents(tx *sql.Tx, blockID, txID uint32, evts []abci.Event) error {
+	// Populate the transaction ID field iff one is defined (> 0).
+	var txIDArg interface{}
+	if txID > 0 {
+		txIDArg = txID
+	}
 
-		sqlStmtTxResult = sqlStmtTxResult.Values(txBz, ts)
+	// Add each event to the events table, and retrieve its row ID to use when
+	// adding any attributes the event provides.
+	for _, evt := range evts {
+		// Skip events with an empty type.
+		if evt.Type == "" {
+			continue
+		}
 
-		// execute sqlStmtTxResult db query and retrieve the txid
-		r, err := sqlStmtTxResult.Query()
+		eid, err := queryWithID(tx, `
+INSERT INTO `+tableEvents+` (block_id, tx_id, type) VALUES ($1, $2, $3)
+  RETURNING rowid;
+`, blockID, txIDArg, evt.Type)
```

Comment:
should we have a file with constants for the queries?

I'm also *deeply suspicious* of anything that does raw string concatenation (because, relative to the `fmt.Sprint<>`, t the lack of buffering has bitten me a few times.)

Reply:
I could pull out the queries, but I didn't because then it's harder for the reader to see what the query does here in context, and this is where it matters. 

The string concatenation happens at compile time, so there's no buffering issue.